### PR TITLE
Post Q2 Tapeout Fixes

### LINKED ISSUE #536: This should say "power rails" not "standard cells"
Happened to find while looking at the slides.

https://github.com/ucb-bar/hammer/blob/85c12e98e63d599f5267c7711a24c174d74974c5/src/hammer-vlsi/technology/asap7/defaults.yml#L54

### LINKED ISSUE #542: Use special cells for the power straps reference cell
We should use "Special cells" to replace the `technology.core.tap_cell_rail_reference` setting.

### LINKED ISSUE #677: Top/bottom routing layers should be new keys
P&R tools (OpenROAD, Innovus) generally require this to be specified for the detailed router. This should be accomplished in conjunction with #675 and the requisite simultaneous plugin updates, which all currently use hooks to set this.

### LINKED ISSUE #784: Don't touch / don't route options
https://github.com/ucb-bar/hammer/blob/0bbd2b5f6cc42c6a07563875d9fb882b26dde8de/hammer/par/innovus/__init__.py#L467

For preplaced pins in macros, this line currently lets Innovus still route the signal to the edge. This is not desired, e.g. if the macro pin is in the middle on a high layer and we want to leave it there.

The fix should be some combination of `set_dont_touch` and `set_db <nets> .skip_routing true`

### LINKED ISSUE #785: Route obstruction options
https://github.com/ucb-bar/hammer-cadence-plugins/blob/902657b8bdba7aec12726d6d3c85a2639a4030ad/hammer/par/innovus/__init__.py#L1014

This line as is obstructs both routing and pg_nets. Is that the desired behavior? 

If we want to obstruct only routing but not pg_nets, we need the flag `except_pg_nets`

Use case: connecting a macro to power grid, but block routing over it to avoid unpredictable coupling effects.

### PR DESCRIPTION
**Pull request by harrisonliew**
Miscellaneous fixes/features post-Q2 tapeout, including:
- Fix ILMs not fully propagating (full hierarchical tree for LVS, see description below)
- Fix macro pin promotion
- Fix routing obstruction type also including PG obstruction
- Add top/bottom routing layer key (due to Innovus command deprecation) & fix some Genus command deprecations
- Use the tapcells list for power rail reference
- Add `prepend` meta directives similar to the append ones. Needed for certain lists where order matters (e.g. include directories)

Details about the LVS ILM fix:
- `vlsi.inputs.ilms` is now propagated through the flow so that each par node appends to it. Thus for any level in the hierarchical tree, it contains the full dependency tree
- `par.outputs.output_ilms` continues to only contain the children ILMs (depth = 1). This is then used for syn/par/timing when `get_input_ilms()` is called so that they don't read extraneous ILMs.
- As a result, `lvs.inputs.ilms` and `timing.inputs.ilms` are unnecessary/extraneous.
- I've added a `mock_hier.yml` config in the e2e tests and new `get_ilms()` steps in mockpar and mocklvs to test this. Simply need to run `hammer-vlsi -p config/mock_hier.yml lvs-ChipTop` to verify.

TODO (optional):
- [ ] Add antenna diodes to special cells list and enable additional antenna fixing settings in Innovus

Technology plugins will need to be updated to work with these changes.

**Related PRs / Issues**

Closes #542, closes #536, closes #677, closes #784, closes #785, closes #654

**Type of change**:
- [X] Bug fix
- [X] New feature
- [ ] Other enhancement

**Impact**:
- [x] Change to core Hammer
- [ ] Change to a Hammer plugin
- [ ] Other

**Contributor Checklist**:
- [x] Did you set `master` as the base branch?
- [x] Did you state the type-of-change/impact?
- [x] Did you delete any extraneous prints/debugging code?
- [ ] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you update the `poetry.lock` file if you updated the requirements in `pyproject.toml`?
- [ ] (If applicable) Did you add a unit test demonstrating the PR?
- [x] (If applicable) Did you run this through the e2e integration tests?
- [ ] (If applicable) Did you update the submodules in `e2e/` if this feature depends on updated plugins?

### CODE REVIEW

File: hammer/par/innovus/__init__.py [hunk 1]
```diff
@@ -249,6 +249,18 @@ def init_design(self) -> bool:
...
+        if layers is not None:
+            if self.version() >= self.version_number("201"):
+                verbose_append(f"set_db design_bottom_layer {layers[0]}")
+                verbose_append(f"set_db design_bottom_layer {layers[1]}")
+            else:
+                verbose_append(f"set_db route_early_global_bottom_layer {layers[0]}")
+                verbose_append(f"set_db route_early_global_bottom_layer {layers[1]}")
+                verbose_append(f"set_db route_design_bottom_layer {layers[0]}")
+                verbose_append(f"set_db route_design_bottom_layer {layers[1]}")
+
```
[the remark this answers is not in the document]
Comment (harrisonliew) [reply hammer/par/innovus/__init__.py:263]: Ah, oops. Will fix tomorrow (or you can go ahead).

**Comment by kenhoberkeley** [comment]
Should we throw the vlsi.inputs.ilms and lvs.inputs.ilms issue into this as well?

**Review by nayiri-k — APPROVED** [review]

**Review by kenhoberkeley — APPROVED** [review]

**Review by jerryz123 — APPROVED** [review]

**Review by moresa75 — APPROVED** [review]

**Comment by boblinchuan** [comment]
https://hammer-vlsi.readthedocs.io/en/stable/Hammer-Use/Hooks.html `test.py` on this page doesn't like to anything...

File: hammer/par/innovus/__init__.py [hunk 2]
```diff
@@ -397,11 +409,21 @@ def place_pins(self) -> bool:
...
+        self.append(f"set_db assign_pins_promoted_macro_top_layer {self.get_stackup().get_metal_by_index(-1).name}")
+        self.append('set all_ppins ""')
         for pin in pin_assignments:
             if pin.preplaced:
-                # First set promoted pins
-                self.verbose_append("set_promoted_macro_pin -pins {{ {p} }}".format(p=pin.pins))
-                promoted_pins.extend(pin.pins.split())
+                # Find the pin object that should be promoted. Only one of the two (driver_pins or load_pins) should be non-empty.
+                self.verbose_append(f'set ppins [get_db [get_nets {pin.pins}] .driver_pins]')
+                self.verbose_append(f'if {{$ppins eq ""}} {{set ppins [get_db [get_nets {pin.pins}] .load_pins]}}')
```
Comment (boblinchuan) [inline hammer/par/innovus/__init__.py:420]: This triggers a double "" from puts, causing Innovus to crash.

**Review by boblinchuan — APPROVED** [review]
Reviewed files relevant to #785 and #784. Tested with my tapeout design. Looks the same, good to go.

**Pull request merged**